### PR TITLE
Update string similarity measures

### PR DESCRIPTION
**Pull request by jstammers**
This PR adds a few commonly used string similarity measures that are built-in to duckdb. 

These are not currently implemented in other backends (e.g. pyspark). It would be useful to be able to dispatch to another UDF based on the backend, e.g.

```python
@ibis.udf.builtin(name="jaro_winkler_similarity")
def jw_duckdb(s1: str, s2: str) -> float:
    ...

@ibis.udf.scalar.python
def jw_python(s1: str, s2: str) -> float:
    # import rapidfuzz and use here

def jaro_winkler(s1: ir.StringValue, s2: ir.StringValue) -> ir.FloatValue:
    backend = ...
    if backend == 'duckdb':
        return jw_duckdb(s1, s2)
    else:
        return jw_python(s1, s2)

```

but that feels a little clunky to me.

I've added the string comparison chart from `splink` because it seems like a convenient way to visualise the differences between the comparison functions. The example notebook could probably do with some more detail, so I'd welcome some comments there.

Finally, I've noticed that the duckdb `jaccard` method doesn't always give the same value as expected from the tokenized set version, even though they seem equivalent. Maybe some more examples are needed to check that these are performing as expected.

### CODE REVIEW

File: mismo/text/_similarity.py [hunk 1]
```diff
@@ -86,7 +94,7 @@ def levenshtein_ratio(s1: ir.StringValue, s2: ir.StringValue) -> ir.FloatingValu
     >>> levenshtein_ratio("", "").execute()
     np.float64(nan)
     """
-    return _dist_ratio(s1, s2, lambda a, b: a.levenshtein(b))
```
Comment (NickCrews) [inline mismo/text/_similarity.py:89]: String.levenshtein() is built into ibis, this is correct as it is.: https://ibis-project.org/reference/expression-strings#ibis.expr.types.strings.StringValue.levenshtein

File: mismo/text/tests/test_similarity.py [hunk 2]
```diff
@@ -38,3 +38,39 @@ def test_levenshtein_ratio(string1, string2, expected):
...
+        assert np.isnan(result)
+    else:
+        assert result == pytest.approx(expected, 0.001)
+
+
+@pytest.mark.parametrize(
+    "string1,string2,expected",
+    [
+        ("foo", "foo", 1),
+        ("foo bar", "foo", 0.3333),  # this is currently failing
```
Comment (NickCrews) [inline mismo/text/tests/test_similarity.py:65]: We would expect this to be 1/2, not 1/3? Or am I not understanding jaccard similarity?

**Comment by NickCrews** [comment]
Thanks for all this! I think it all looks like the right idea. However, I don't want to have to think about all these changes in the same PR. Can you make one PR just for jaro winkler stuff? I think that looks pretty much good to go. Then a separate PR for the jaccard(str, str) function? I have a few thoughts there, and that will slow down merging the jaro stuff. Same for the charting stuff, I think that's a good idea, but please do it in a separate PR so it's not slowing down the other changes.

> It would be useful to be able to dispatch to another UDF based on the backend

I've run into this problem too. I think your solution is the best we can do at the moment. See https://github.com/ibis-project/ibis/issues/8748

**Review by NickCrews — CHANGES_REQUESTED** [review]

**Comment by jstammers** [comment]
Thanks for the feedback @NickCrews. I've made the relevant suggestions and have split the jaccard function and comparison utilities into other PRs

File: mismo/text/__init__.py [hunk 3]
```diff
@@ -3,10 +3,13 @@
 from __future__ import annotations
 
 from mismo.text._features import ngrams as ngrams
+from mismo.text._features import tokenize as tokenize
 from mismo.text._similarity import damerau_levenshtein as damerau_levenshtein
 from mismo.text._similarity import (
     damerau_levenshtein_ratio as damerau_levenshtein_ratio,
 )
 from mismo.text._similarity import double_metaphone as double_metaphone
+from mismo.text._similarity import jaro_similarity as jaro_similarity
```
Comment (NickCrews) [inline mismo/text/__init__.py:12]: Please add these to `docs/reference/text.md`

**Comment by NickCrews** [comment]
Sweet, looks good! Once you add these to the docs it looks good to go! Thank you!

**Review by jstammers — COMMENTED** [review]
Added those to the docs now

**Pull request merged**